File: 8. String to Integer (atoi).py
class Solution:
    def myAtoi(self, str: str) -> int:
        '''
        remember to do str.lstrip() first, this is to reduce the leading ' '.
        '''
        str = str.lstrip()
        if not str:
            return 0
        sign = 1
        i = 0
        if str[0] == '+' or str[0] == '-':
            if len(str) == 1:
                return 0
            else:
                sign = 1 if str[0] == '+' else -1
                str = str[1:]
        while i < len(str) and str[i].isdigit():
            i += 1
        ret = sign * int(str[0:i] if i > 0 else 0)
        if ret < -2 ** 31:
            return -2 ** 31
        elif ret > 2**31 - 1:
            return 2**31 - 1
        else:
            return ret

File: 8. String to Integer (atoi).java
class Solution {
    public int myAtoi(String str) {
        str = str.trim();
        int result = 0;
        int sign = 1;
        for (int i=0; i<str.length(); i++) {
            if (str.charAt(i)=='+' || str.charAt(i)=='-') {
                if (i!=0) {
                    return sign * result;
                } else {
                    sign = str.charAt(i)=='+' ? 1 : -1;
                }
            } else if (Character.isDigit(str.charAt(i))) {
                int temp = str.charAt(i) - '0';
                if (result * sign > Integer.MAX_VALUE/10 || (result * sign == Integer.MAX_VALUE/10 && temp > 7)) {
                    return Integer.MAX_VALUE;
                } else if (result * sign < Integer.MIN_VALUE/10 || (result * sign == Integer.MIN_VALUE/10 && temp > 8)) {
                    return Integer.MIN_VALUE;
                } else {
                    result = result * 10 + temp;
                }
            } else {
                return sign * result;
            }
        }
        return sign * result;
    }
}